### PR TITLE
Utils: Add string padding and hex formatting to TEString

### LINKED ISSUE #193: Add string padding, character occurrence counting, and hex formatting to TEString
### 📝 Description
`TEString` is TimeEngine's custom string class. To enhance string formatting capabilities for serialization, debug logging, and data inspection, we want to expand `TEString` with padding, character frequency counting, and hexadecimal formatting methods.

> **Note:** The functions listed below represent the baseline requirements and examples. Contributors are welcome to suggest or implement additional string utilities that would be helpful for engine-wide formatting and text manipulation!

---

### 🎯 Scope & Expected Tasks

Modify `Engine/Include/Utils/TEString.hpp` and `Engine/src/Utils/TEString.cpp` to implement the following helpers (along with any other useful string utilities):

#### 1. Padding & Inspection
- **`TEString PadLeft(size_t totalLength, char paddingChar = ' ') const`**
  - Pads the left side of the string with `paddingChar` until the total string length reaches `totalLength`.
  - *Edge case*: If `totalLength` is less than or equal to the current string length, return a copy of the original string without modifying or truncating.
- **`TEString PadRight(size_t totalLength, char paddingChar = ' ') const`**
  - Pads the right side of the string with `paddingChar` until the total string length reaches `totalLength`.
  - *Edge case*: If `totalLength` is less than or equal to the current string length, return a copy of the original string without modifying or truncating.
- **`int CountOccurrences(char ch) const`**
  - Counts and returns the number of times `ch` appears in the string.

#### 2. Hex Formatting
- **`static TEString ToHex(uint32_t value, bool uppercase = false, bool prefix = false)`**
  - Converts a 32-bit unsigned integer to its hexadecimal string representation.
  - Supports configurable casing (`uppercase = true` $\rightarrow$ `"FF"`, `uppercase = false` $\rightarrow$ `"ff"`).
  - Supports optional `0x` / `0X` prefixing when `prefix = true`.
- **`static TEString ToHex64(uint64_t value, bool uppercase = false, bool prefix = false)`**
  - 64-bit integer overload providing the same formatting options as `ToHex`.

---

### 💡 General Guidance
- Member methods marked `const` must not modify the original instance and should return a new `TEString`.
- Follow established `TimeEngine` naming conventions and internal buffer management patterns.

---

### 📂 Relevant Files
- `Engine/Include/Utils/TEString.hpp`
- `Engine/src/Utils/TEString.cpp`

---

### 🧪 Verification Steps
1. **Behavioral Checks**:
   - Verify `PadLeft("42", 5, '0')` yields `"00042"`.
   - Verify `PadLeft("Hello", 3, ' ')` returns `"Hello"` without truncation.
   - Verify `ToHex(255, true, true)` yields `"0xFF"` and `ToHex(255, false, false)` yields `"ff"`.
   - Verify `CountOccurrences('a')` accurately counts matching characters across empty and non-empty strings.
2. **Build Verification**:
   - Compile the project using your platform's relevant build script located under `Scripts/`.

### PR DESCRIPTION
**Pull request by Aditya-233**
## Description

Expands `TEString` with string padding, in-place mutating variants, frequency counting, and stack-buffered hexadecimal formatting helpers.

1. **Left & Right Padding (`PadLeft` / `PadRight`):**
    - `PadLeft(size_t totalLength, const TEString &padding) const`
    - `PadLeft(size_t totalLength, char paddingChar = ' ') const`
    - `PadRight(size_t totalLength, const TEString &padding) const`
    - `PadRight(size_t totalLength, char paddingChar = ' ') const`
    - _Design details:_ Pre-calculates exact full repeats (`padLen / patLen`) and remainder (`padLen % patLen`) with `.reserve()`, preventing buffer overallocation and eliminating `resize()` calls. Single-character padding uses direct `std::string` fill methods with zero temporary heap object construction.

2. **In-Place Transformation Counterparts (`*Inline`):**
    - `void PadLeftInline(size_t totalLength, const TEString &padding)`
    - `void PadLeftInline(size_t totalLength, char paddingChar = ' ')`
    - `void PadRightInline(size_t totalLength, const TEString &padding)`
    - `void PadRightInline(size_t totalLength, char paddingChar = ' ')`
    - Matches the engine's core dual-API pattern (`Trim()` / `TrimInline()`, `Reverse()` / `ReverseInline()`).

3. **Character & Substring Frequency Counting (`CountOccurrences`):**
    - `size_t CountOccurrences(const TEString &sub) const`
    - `size_t CountOccurrences(char ch) const`
    - Returns `size_t` to match string length conventions and eliminate signed/unsigned comparison warnings. Single-character counting utilizes `std::count`.

4. **Hexadecimal Formatting (`ToHex` / `ToHex64`):**
    - `static TEString ToHex(uint32_t value, bool uppercase = false, bool prefix = false)`
    - `static TEString ToHex64(uint64_t value, bool uppercase = false, bool prefix = false)`
    - Replaced `std::ostringstream` with `std::snprintf` using stack buffers (`char buf[32]` and `char buf[48]`), eliminating heap allocations in hot paths.

---

## Related Issues

Fixes #193

---

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Style / Refactoring

---

## Build Environment & OS

- **Operating System:** Arch Linux (`Linux 7.1.9-arch1-2 x86_64`)
- **Compiler / Toolchain:** `g++ (GCC) 16.2.1 20260810`
- **C++ Standard:** C++20 (`-std=c++20 -Wall -Wextra -Werror`)

---

## How Has This Been Tested?

Compiled and executed directly against `Engine/src/Utils/TEString.cpp` and `Engine/src/Core/PreRequisites.cpp` with strict flags:

```bash
g++ -std=c++20 -Wall -Wextra -Werror \
  -IEngine/Include -IEngine/src \
  Engine/src/Utils/TEString.cpp \
  Engine/src/Core/PreRequisites.cpp \
  -x c++ - -o /tmp/timeengine_verification << 'EOF'
#include <iostream>
#include <cassert>
#include <cstdint>
#include "Utils/TEString.hpp"

int main() {
    std::cout << "Starting TimeEngine TEString verification test suite..." << std::endl;

    // 1. Single character with default and explicit padding (returning & inline)
    {
        TEString s("42");
        assert(s.PadLeft(5) == "   42");
        assert(s.PadRight(5) == "42   ");
        assert(s.PadLeft(5, '0') == "00042");
        assert(s.PadRight(5, '0') == "42000");

        // Edge case: totalLength <= length
        assert(s.PadLeft(2, '0') == "42");
        assert(s.PadLeft(1, '0') == "42");
        assert(s.PadRight(2, '0') == "42");

        // In-place mutators
        TEString sInline1("42");
        sInline1.PadLeftInline(5);
        assert(sInline1 == "   42");

        TEString sInline2("42");
        sInline2.PadLeftInline(5, '0');
        assert(sInline2 == "00042");

        TEString sInline3("42");
        sInline3.PadRightInline(5);
        assert(sInline3 == "42   ");

        TEString sInline4("42");
        sInline4.PadRightInline(5, '0');
        assert(sInline4 == "42000");
    }

    // 2. Multi-character patterns and fractional clipping (returning & inline)
    {
        TEString s("42");
        // Exact multiple repeats (padLen = 6, patLen = 3)
        assert(s.PadLeft(8, TEString("abc")) == "abcabc42");
        assert(s.PadRight(8, TEString("abc")) == "42abcabc");

        // Fractional remainder clipping (padLen = 5, patLen = 3 -> 1 repeat + 2 chars)
        assert(s.PadLeft(7, TEString("abc")) == "abcab42");
        assert(s.PadRight(7, TEString("abc")) == "42abcab");

        // Empty pattern & empty source string edge cases
        assert(s.PadLeft(10, TEString("")) == "42");
        assert(s.PadRight(10, TEString("")) == "42");

        TEString empty("");
        assert(empty.PadLeft(5, TEString("ab")) == "ababa");
        assert(empty.PadRight(5, TEString("ab")) == "ababa");

        // In-place pattern mutation
        TEString sPatLeft("42");
        sPatLeft.PadLeftInline(7, TEString("abc"));
        assert(sPatLeft == "abcab42");

        TEString sPatRight("42");
        sPatRight.PadRightInline(7, TEString("abc"));
        assert(sPatRight == "42abcab");
    }

    // 3. Character and Substring Frequency Counting (CountOccurrences)
    {
        TEString s("banana");
        // Substrings (const TEString&)
        assert(s.CountOccurrences(TEString("na")) == 2);
        assert(s.CountOccurrences(TEString("a")) == 3);
        assert(s.CountOccurrences(TEString("ban")) == 1);
        assert(s.CountOccurrences(TEString("xyz")) == 0);
        assert(s.CountOccurrences(TEString("")) == 0);

        // Characters (char)
        assert(s.CountOccurrences('a') == 3);
        assert(s.CountOccurrences('n') == 2);
        assert(s.CountOccurrences('b') == 1);
        assert(s.CountOccurrences('z') == 0);

        TEString empty("");
        assert(empty.CountOccurrences('a') == 0);
        assert(empty.CountOccurrences(TEString("a")) == 0);
    }

    // 4. Hexadecimal Formatting (ToHex & ToHex64)
    {
        // 32-bit values, casing, and prefixing
        assert(TEString::ToHex(255, true, true) == "0xFF");
        assert(TEString::ToHex(255, false, false) == "ff");
        assert(TEString::ToHex(255, true, false) == "FF");
        assert(TEString::ToHex(255, false, true) == "0xff");

        // 32-bit extremes (0 and UINT32_MAX)
        assert(TEString::ToHex(0, true, true) == "0x0");
        assert(TEString::ToHex(0, false, false) == "0");
        assert(TEString::ToHex(UINT32_MAX, true, true) == "0xFFFFFFFF");
        assert(TEString::ToHex(UINT32_MAX, false, false) == "ffffffff");

        // 64-bit values & extremes (0 and UINT64_MAX)
        assert(TEString::ToHex64(0x123456789ABCDEF0ULL, true, true) == "0x123456789ABCDEF0");
        assert(TEString::ToHex64(0x123456789ABCDEF0ULL, false, false) == "123456789abcdef0");
        assert(TEString::ToHex64(0, true, true) == "0x0");
        assert(TEString::ToHex64(0, false, false) == "0");
        assert(TEString::ToHex64(UINT64_MAX, true, true) == "0xFFFFFFFFFFFFFFFF");
        assert(TEString::ToHex64(UINT64_MAX, false, false) == "ffffffffffffffff");
    }

    std::cout << "ALL TEST ASSERTIONS PASSED (100% SUCCESS, 0 ERRORS, 0 WARNINGS)" << std::endl;
    return 0;
}
EOF
/tmp/timeengine_verification
```

### Execution Output:

```
Starting TimeEngine TEString verification test suite...
ALL TEST ASSERTIONS PASSED (100% SUCCESS, 0 ERRORS, 0 WARNINGS)
```

---

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] **Determinism Check**: If adding logic, I have ensured it is deterministic and time-aware.

### CODE REVIEW

**Comment by 1SHAMAY1** [comment]
@Aditya-233 please follow the commit and PR guidelines

**Comment by 1SHAMAY1** [comment]
@Aditya-233 Listen we currently don't have proper deterministic environment.
So I clearly do not understand how you tested it and marked it done.

File: Engine/Include/Utils/TEString.hpp [hunk 1]
```diff
@@ -192,6 +192,10 @@ class TE_API TEString
     void TrimEndInline();
     void TrimInline();
 
+    TEString PadLeft(size_t totalLength, char paddingChar = ' ') const;
```
Comment (1SHAMAY1) [inline Engine/Include/Utils/TEString.hpp:196]: Please use `TESTRING` as an input.

**Review by 1SHAMAY1 — COMMENTED** [review]
@Aditya-233 here is a review for your code

File: Engine/Include/Utils/TEString.hpp [hunk 2]
```diff
@@ -192,6 +192,10 @@ class TE_API TEString
...
+    TEString PadRight(size_t totalLength, char paddingChar = ' ') const;
```
Comment (1SHAMAY1) [inline Engine/Include/Utils/TEString.hpp:201]: Here as well

**Review by 1SHAMAY1 — CHANGES_REQUESTED** [review]
Here do same as the previous review

**Comment by Aditya-233** [comment]
> @Aditya-233 Listen we currently don't have proper deterministic environment. So I clearly do not understand how you tested it and marked it done.

Because `TEString` only depends on `Core/PreRequisites.h`, we can easily compile and test all functions (`PadLeft`, `PadRight`, `CountOccurrences`, `ToHex`, `ToHex64`) without needing the full game engine to boot up.

Here is the exact command and test script that was executed locally:

```bash
g++ -std=c++20 -Wall -Wextra -Werror \
  -IEngine/Include -IEngine/src \
  Engine/src/Utils/TEString.cpp \
  Engine/src/Core/PreRequisites.cpp \
  -x c++ - << 'EOF'
#include <iostream>
#include <cassert>
#include "Utils/TEString.hpp"

int main() {
    // 1. Test PadLeft
    TEString s1("42");
    assert(s1.PadLeft(5, '0') == "00042");
    assert(s1.PadLeft(2, '0') == "42");
    assert(s1.PadLeft(1, '0') == "42");
    TEString sHello("Hello");
    assert(sHello.PadLeft(3, ' ') == "Hello");

    // 2. Test PadRight
    TEString s2("42");
    assert(s2.PadRight(5, '0') == "42000");
    assert(s2.PadRight(2, '0') == "42");
    assert(s2.PadRight(1, '0') == "42");
    assert(sHello.PadRight(3, ' ') == "Hello");

    // 3. Test CountOccurrences
    TEString sBanana("banana");
    assert(sBanana.CountOccurrences('a') == 3);
    assert(sBanana.CountOccurrences('b') == 1);
    assert(sBanana.CountOccurrences('z') == 0);
    TEString sEmpty("");
    assert(sEmpty.CountOccurrences('a') == 0);

    // 4. Test ToHex
    assert(TEString::ToHex(255, true, true) == "0xFF");
    assert(TEString::ToHex(255, false, false) == "ff");
    assert(TEString::ToHex(255, true, false) == "FF");
    assert(TEString::ToHex(255, false, true) == "0xff");
    assert(TEString::ToHex(0, false, true) == "0x0");

    // 5. Test ToHex64
    assert(TEString::ToHex64(0x123456789ABCDEF0ULL, true, true) == "0x123456789ABCDEF0");
    assert(TEString::ToHex64(0x123456789abcdef0ULL, false, false) == "123456789abcdef0");
    assert(TEString::ToHex64(0, false, false) == "0");

    std::cout << "All TEString unit tests passed successfully!" << std::endl;
    return 0;
}
EOF
```

**Comment by Aditya-233** [comment]
Thanks for the feedback, i looked through it again
How does this look??
```diff
diff --git a/Engine/Include/Utils/TEString.hpp b/Engine/Include/Utils/TEString.hpp
index 35c788f..4096f1d 100644
--- a/Engine/Include/Utils/TEString.hpp
+++ b/Engine/Include/Utils/TEString.hpp
@@ -192,8 +192,11 @@ public:
     void TrimEndInline();
     void TrimInline();

-    TEString PadLeft(size_t totalLength, char paddingChar = ' ') const;
-    TEString PadRight(size_t totalLength, char paddingChar = ' ') const;
+    TEString PadLeft(size_t totalLength, const TEString &padding = " ") const;
+    TEString PadLeft(size_t totalLength, char paddingChar) const;
+    TEString PadRight(size_t totalLength, const TEString &padding = " ") const;
+    TEString PadRight(size_t totalLength, char paddingChar) const;
+    int CountOccurrences(const TEString &sub) const;
     int CountOccurrences(char ch) const;

     TEString Replace(const TEString &from, const TEString &to,
diff --git a/Engine/src/Utils/TEString.cpp b/Engine/src/Utils/TEString.cpp
index fa1194c..8061178 100644
--- a/Engine/src/Utils/TEString.cpp
+++ b/Engine/src/Utils/TEString.cpp
@@ -469,26 +469,71 @@ void TEString::TrimEndInline() { *this = TrimEnd(); }

 void TEString::TrimInline() { *this = Trim(); }

-TEString TEString::PadLeft(size_t totalLength, char paddingChar) const
+TEString TEString::PadLeft(size_t totalLength, const TEString &padding) const
 {
-    if (totalLength <= m_Data.length())
+    if (totalLength <= m_Data.length() || padding.IsEmpty())
         return *this;

-    std::string result(totalLength - m_Data.length(), paddingChar);
-    result += m_Data;
-    return TEString(std::move(result));
+    size_t padLen = totalLength - m_Data.length();
+    std::string padStr;
+    padStr.reserve(padLen);
+    while (padStr.length() < padLen)
+    {
+        padStr.append(padding.m_Data);
+    }
+    if (padStr.length() > padLen)
+    {
+        padStr.resize(padLen);
+    }
+
+    padStr += m_Data;
+    return TEString(std::move(padStr));
 }

-TEString TEString::PadRight(size_t totalLength, char paddingChar) const
+TEString TEString::PadLeft(size_t totalLength, char paddingChar) const
+{
+    return PadLeft(totalLength, TEString(&paddingChar, 1));
+}
+
+TEString TEString::PadRight(size_t totalLength, const TEString &padding) const
 {
-    if (totalLength <= m_Data.length())
+    if (totalLength <= m_Data.length() || padding.IsEmpty())
         return *this;

     std::string result = m_Data;
-    result.append(totalLength - m_Data.length(), paddingChar);
+    result.reserve(totalLength);
+    while (result.length() < totalLength)
+    {
+        result.append(padding.m_Data);
+    }
+    if (result.length() > totalLength)
+    {
+        result.resize(totalLength);
+    }
+
     return TEString(std::move(result));
 }

+TEString TEString::PadRight(size_t totalLength, char paddingChar) const
+{
+    return PadRight(totalLength, TEString(&paddingChar, 1));
+}
+
+int TEString::CountOccurrences(const TEString &sub) const
+{
+    if (sub.IsEmpty() || m_Data.empty())
+        return 0;
+
+    int count = 0;
+    size_t pos = 0;
+    while ((pos = m_Data.find(sub.m_Data, pos)) != std::string::npos)
+    {
+        ++count;
+        pos += sub.m_Data.length();
+    }
+    return count;
+}
+
 int TEString::CountOccurrences(char ch) const
 {
     int count = 0;
```


and this is the new test i did
```bash
g++ -std=c++20 -Wall -Wextra -Werror \
  -IEngine/Include -IEngine/src \
  Engine/src/Utils/TEString.cpp \
  Engine/src/Core/PreRequisites.cpp \
  -x c++ - << 'EOF'
#include <iostream>
#include <cassert>
#include "Utils/TEString.hpp"

int main() {
    // 1. PadLeft tests
    TEString s1("42");
    assert(s1.PadLeft(5) == "   42");
    assert(s1.PadLeft(5, TEString("0")) == "00042");
    assert(s1.PadLeft(5, "0") == "00042");
    assert(s1.PadLeft(7, TEString("abc")) == "abcab42");
    assert(s1.PadLeft(5, '0') == "00042");
    assert(s1.PadLeft(2, TEString("0")) == "42");
    assert(s1.PadLeft(5, TEString("")) == "42");

    // 2. PadRight tests
    TEString s2("42");
    assert(s2.PadRight(5) == "42   ");
    assert(s2.PadRight(5, TEString("0")) == "42000");
    assert(s2.PadRight(5, "0") == "42000");
    assert(s2.PadRight(7, TEString("abc")) == "42abcab");
    assert(s2.PadRight(5, '0') == "42000");
    assert(s2.PadRight(2, TEString("0")) == "42");
    assert(s2.PadRight(5, TEString("")) == "42");

    // 3. CountOccurrences tests
    TEString sBanana("banana");
    assert(sBanana.CountOccurrences(TEString("na")) == 2);
    assert(sBanana.CountOccurrences("na") == 2);
    assert(sBanana.CountOccurrences(TEString("a")) == 3);
    assert(sBanana.CountOccurrences(TEString("ban")) == 1);
    assert(sBanana.CountOccurrences(TEString("xyz")) == 0);
    assert(sBanana.CountOccurrences(TEString("")) == 0);
    assert(sBanana.CountOccurrences('a') == 3);

    // 4. ToHex and ToHex64 tests
    assert(TEString::ToHex(255, true, true) == "0xFF");
    assert(TEString::ToHex(255, false, false) == "ff");
    assert(TEString::ToHex64(0x123456789ABCDEF0ULL, true, true) == "0x123456789ABCDEF0");

    std::cout << "TESTRING DIFF TEST SUITE: 100% SUCCESS!" << std::endl;
    return 0;
}
EOF
```

**Comment by 1SHAMAY1** [comment]
@Aditya-233 I agree with your local test cases and the way you verified the functions. However, leaving the Determinism Check box ticked in the PR description is a bit misleading. As you noted, we don't currently have a proper deterministic testing environment, so please uncheck that box to keep the PR status accurate.
​Also, please make sure to check the inline code review and address the requested changes (specifically, using TEString as an input for the padding methods). Let me know when you have updated the branch!

**Comment by Aditya-233** [comment]
Thanks for the clarification! I've updated the PR:

1. **Unchecked Determinism Box**: Updated the PR description checklist to leave `[ ] **Determinism Check**` unchecked to accurately reflect current environment status.
2. **Addressed Code Review Feedback**:
   - Updated `PadLeft` and `PadRight` to accept `const TEString &padding = " "` as requested, with overloaded `char` literals for convenience.
   - Updated `CountOccurrences` to support `const TEString &sub` for substring counting alongside `char ch`.
3. **Local Test Verification**:
   - Re-verified all assertions (`PadLeft`, `PadRight`, `CountOccurrences`, `ToHex`, `ToHex64`) with `g++ -std=c++20 -Wall -Wextra -Werror` (100% pass, 0 warnings).
Any suggestions maybe??

**Comment by Aditya-233** [comment]
Hi @1SHAMAY1, can you review and suggest any changes if needed?

**Comment by 1SHAMAY1** [comment]
@Aditya-233 Thanks for getting those updates in! I’ll be doing a deeper review of the code over the next few days to make sure everything aligns perfectly with our architecture, but I plan to merge this shortly once that's wrapped up.
In the meantime, definitely take a look at the project roadmap. It would also be a huge help if you could spend some time testing the engine itself—let me know if you run into any weird behaviors or if there are any specific utilities or features you feel are missing.
By the way, what specific OS and build environment do you currently have?

File: Engine/src/Utils/TEString.cpp [hunk 3]
```diff
@@ -469,6 +469,82 @@ void TEString::TrimEndInline() { *this = TrimEnd(); }
 
 void TEString::TrimInline() { *this = Trim(); }
 
+TEString TEString::PadLeft(size_t totalLength, const TEString &padding) const
+{
```
Comment (1SHAMAY1) [inline Engine/src/Utils/TEString.cpp:473]: Pre-calculating fullRepeats and remainder prevents overshooting the padding length and eliminates the extra `resize()` call. Also, reserving totalLength avoids reallocating when appending `m_Data`.

File: Engine/src/Utils/TEString.cpp [hunk 4]
```diff
@@ -469,6 +469,82 @@ void TEString::TrimEndInline() { *this = TrimEnd(); }
...
+    if (padStr.length() > padLen)
+    {
+        padStr.resize(padLen);
+    }
+
+    padStr += m_Data;
+    return TEString(std::move(padStr));
+}
+
+TEString TEString::PadLeft(size_t totalLength, char paddingChar) const
```
Comment (1SHAMAY1) [inline Engine/src/Utils/TEString.cpp:496]: Avoid delegating through a temporary `TEString(&paddingChar, 1)` object here. Implement the `char` overload directly with `std::string` fill/append methods so single-character padding is zero-overhead.

File: Engine/Include/Utils/TEString.hpp [hunk 5]
```diff
@@ -192,6 +192,13 @@ class TE_API TEString
     void TrimEndInline();
     void TrimInline();
 
+    TEString PadLeft(size_t totalLength, const TEString &padding = " ") const;
+    TEString PadLeft(size_t totalLength, char paddingChar) const;
+    TEString PadRight(size_t totalLength, const TEString &padding = " ") const;
+    TEString PadRight(size_t totalLength, char paddingChar) const;
+    int CountOccurrences(const TEString &sub) const;
```
Comment (1SHAMAY1) [inline Engine/Include/Utils/TEString.hpp]: Please change the return type from `int` to `size_t (or uint32_t)` to match string size/length conventions in `TEString` and avoid signed/unsigned comparison warnings.

File: Engine/Include/Utils/TEString.hpp [hunk 6]
```diff
@@ -192,6 +192,13 @@ class TE_API TEString
     void TrimEndInline();
     void TrimInline();
 
+    TEString PadLeft(size_t totalLength, const TEString &padding = " ") const;
```
Comment (1SHAMAY1) [inline Engine/Include/Utils/TEString.hpp]: Supporting both `TEString` (engine type) and char is great! However, let's put the default parameter = ' ' on the char overload rather than `const TEString &padding = " "`. This avoids constructing a temporary `TEString` when someone calls `PadLeft(10)` with default spacing.

File: Engine/src/Utils/TEString.cpp [hunk 7]
```diff
@@ -469,6 +469,82 @@ void TEString::TrimEndInline() { *this = TrimEnd(); }
...
+{
+    return PadLeft(totalLength, TEString(&paddingChar, 1));
+}
+
+TEString TEString::PadRight(size_t totalLength, const TEString &padding) const
```
Comment (1SHAMAY1) [inline Engine/src/Utils/TEString.cpp:513]: Instead of looping with `append` and then calling `resize()` after overshooting, calculate `padLen / padding.Len()` (repeats) and `padLen % padding.Len()` (remainder) directly to build the exact length without extra reallocations.

File: Engine/src/Utils/TEString.cpp [hunk 8]
```diff
@@ -469,6 +469,82 @@ void TEString::TrimEndInline() { *this = TrimEnd(); }
...
+    }
+    if (result.length() > totalLength)
+    {
+        result.resize(totalLength);
+    }
+
+    return TEString(std::move(result));
+}
+
+TEString TEString::PadRight(size_t totalLength, char paddingChar) const
```
Comment (1SHAMAY1) [inline Engine/src/Utils/TEString.cpp:537]: Same here: implement this directly on `m_Data` without constructing a temporary `TEString`.

File: Engine/src/Utils/TEString.cpp [hunk 9]
```diff
@@ -469,6 +469,82 @@ void TEString::TrimEndInline() { *this = TrimEnd(); }
...
+{
+    return PadRight(totalLength, TEString(&paddingChar, 1));
+}
+
+int TEString::CountOccurrences(const TEString &sub) const
```
Comment (1SHAMAY1) [inline Engine/src/Utils/TEString.cpp]: Update the return type and internal counter to `size_t` to match `TEString` length conventions and avoid signed/unsigned comparison warnings.

File: Engine/src/Utils/TEString.cpp [hunk 10]
```diff
@@ -469,6 +469,82 @@ void TEString::TrimEndInline() { *this = TrimEnd(); }
...
+    size_t pos = 0;
+    while ((pos = m_Data.find(sub.m_Data, pos)) != std::string::npos)
+    {
+        ++count;
+        pos += sub.m_Data.length();
+    }
+    return count;
+}
+
+int TEString::CountOccurrences(char ch) const
```
Comment (1SHAMAY1) [inline Engine/src/Utils/TEString.cpp]: Update return type to `size_t`. You can also simplify this loop using standard algorithms like `std::count`.

File: Engine/src/Utils/TEString.cpp [hunk 11]
```diff
@@ -668,6 +744,30 @@ TEString TEString::FromVector4(const TEVector4 &v)
     return TEString::Printf("X=%.3f Y=%.3f Z=%.3f W=%.3f", v.x, v.y, v.z, v.w);
 }
 
+TEString TEString::ToHex(uint32_t value, bool uppercase, bool prefix)
```
Comment (1SHAMAY1) [inline Engine/src/Utils/TEString.cpp:773]: `std::ostringstream` is too heavy and causes unnecessary heap allocations for simple hex formatting in engine code. Please rewrite these using `snprintf` or `TEString::Printf` with stack buffers.

File: Engine/src/Utils/TEString.cpp [hunk 12]
```diff
@@ -668,6 +744,30 @@ TEString TEString::FromVector4(const TEVector4 &v)
...
+    if (uppercase)
+        ss << std::uppercase;
+    ss << std::hex << value;
+    std::string hexStr = ss.str();
+    if (prefix)
+        return TEString("0x" + hexStr);
+    return TEString(std::move(hexStr));
+}
+
+TEString TEString::ToHex64(uint64_t value, bool uppercase, bool prefix)
```
Comment (1SHAMAY1) [inline Engine/src/Utils/TEString.cpp:781]: `std::ostringstream` is too heavy and causes unnecessary heap allocations for simple hex formatting in engine code. Please rewrite these using `snprintf` or `TEString::Printf` with stack buffers.

**Review by 1SHAMAY1 — CHANGES_REQUESTED** [review]
@Aditya-233 Here is another round of review.

**Comment by 1SHAMAY1** [comment]
@Aditya-233 
Thanks for providing your local testing command!

While this verifies basic single-character happy paths, your test script currently has several critical blind spots:

1. **String/Pattern Overloads are Untested:**
   - In sections 1 & 2, all tests pass single `char`s (`'0'`, `' '`). The `const TEString &padding` overloads (multi-character patterns like `s.PadLeft(7, TEString("abc"))`) are completely untested.
   - In section 3, `CountOccurrences` is only tested with `char`, omitting substring searches (`CountOccurrences(const TEString&)`).

2. **Missing Edge Cases:**
   - Empty inputs: `PadLeft(10, TEString(""))`, `CountOccurrences(TEString(""))`.
   - String clipping: test cases where `(totalLength - len)` is not an exact multiple of `padding.Len()`.
   - Extremes: `ToHex64` with `UINT64_MAX` and `ToHex(0, true, true)`.

Please expand your test script to cover these pattern and edge-case scenarios alongside the requested optimizations.

**Comment by 1SHAMAY1** [comment]
@Aditya-233 One additional design note for engine consistency:

`TEString` follows a pattern where every transformation provides both a returning version and an in-place version (e.g., `Trim()` / `TrimInline()`, `Replace()` / `ReplaceInline()`, `Reverse()` / `ReverseInline()`).

Consider also adding the matching `*Inline` methods for the new padding functions to stay 100% consistent with the rest of `TEString`.

**Comment by 1SHAMAY1** [comment]
@Aditya-233 Thanks for the updates!
​Moving forward, please ensure your PRs either use atomic commits (one commit per feature) or are squashed into a single commit. For this specific issue, let's just stick with a single commit.
​Before we move forward, could you update a few things:

- Update the PR description's "How Has This Been Tested?" section with your exact test script and the output logs. ​Add your current build environment and OS details to the PR description.

​I will review the code changes more thoroughly by tomorrow, and we can get this merged shortly after!

**Comment by 1SHAMAY1** [comment]
@Aditya-233 Please rebase and force push, your branch is out of date.

**Review by 1SHAMAY1 — APPROVED** [review]

**Comment by 1SHAMAY1** [comment]
@Aditya-233 Please follow our commit guidelines and squash these commits into one.

**Comment by Aditya-233** [comment]
@1SHAMAY1 Rebased onto the latest upstream/main, squashed into a single atomic commit, and formatted with clang-format as requested.

**Review by 1SHAMAY1 — APPROVED** [review]

**Review by 1SHAMAY1 — APPROVED** [review]

**Comment by 1SHAMAY1** [comment]
Hey, merged your PR into main. Thanks for the contribution! 🎉

Pull the latest main when you get a chance.

Also, if you have some time, feel free to:

Check out any open issues and see if anything interests you
Take a look at the roadmap tracked issues — there's some good stuff in there that could use attention.
Pick up anything else you feel comfortable tackling
No pressure, but always great to have you involved! 🙌
